### PR TITLE
Add support for MinGW-w64

### PR DESCRIPTION
**Pull request by raedrizqie**
MinGW-w64  support

### CODE REVIEW

**Comment by mhammond** [comment]
Is there any chance of submitting the scintilla patches to the scintilla project? Then I can just update it rather than forking it.

**Comment by raedrizqie** [comment]
Ok.. I'll try

**Comment by raedrizqie** [comment]
> 
> 
> Is there any chance of submitting the scintilla patches to the scintilla project? Then I can just update it rather than forking it.

done..

https://github.com/mirror/scintilla/pull/10

**Comment by mhammond** [comment]
That issue is just a single makefile - I was expecting to see most of the .cxx files here in that patch?

**Comment by raedrizqie** [comment]
I've try compiling them without any mods, successfully. Pythonwin only needs the scintilla dll (without lexer) right?

**Comment by mhammond** [comment]
> Pythonwin only needs the scintilla dll (without lexer) right?

pywin32 builds it from source - which is why the .cxx files exist in pwin32.

**Comment by raedrizqie** [comment]
Meaning, you want me to also update the scintilla folder in pywin32 repo?

**Comment by mhammond** [comment]
> Meaning, you want me to also update the scintilla folder in pywin32 repo?

The pythonwin/scintilla directory in pywin32 is a clone of scintilla itself. So all changes in your patch in this issue which are in that directory should rather be submitted to scintilla. However, the PR you opened in scintilla doesn't seem to have any of those files.

**Comment by raedrizqie** [comment]
Done.. https://github.com/raedrizqie/scintilla/commit/774aa40c39ccd0b34acd34f22ec2e8f6054ba76c

I also updated my repo.

File: win32/src/win32rasmodule.cpp [hunk 1]
```diff
@@ -715,7 +715,7 @@ static PyObject *PyRasGetErrorString(PyObject *self, PyObject *args)
 
     TCHAR buf[512];
     // @pyseeapi RasGetErrorString
-    if (rc = RasGetErrorString(error, buf, sizeof(buf) / sizeof(buf[0])))
+    if (rc == RasGetErrorString(error, buf, sizeof(buf) / sizeof(buf[0])))
```
Comment (mhammond) [inline win32/src/win32rasmodule.cpp:718]: All of these changes look wrong - the intent is that rc be passed to `ReturnRasError` - IOW, this block should read:

```
    if (rc = RasGetErrorString(error, buf, sizeof(buf) / sizeof(buf[0])))
        return ReturnRasError("RasGetErrorString", rc);
```

(it's not clear that's actually necessary as `ReturnRasError()` will call `GetLastError()` when we don't pass the error code, but it's probably safer to not rely on `GetLastError()` when we don't need to)

File: Pythonwin/Scintilla/License.txt [hunk 2]
```diff
@@ -2,19 +2,19 @@ License for Scintilla and SciTE
 
 Copyright 1998-2003 by Neil Hodgson <neilh@scintilla.org>
 
-All Rights Reserved 
+All Rights Reserved
```
Comment (mhammond) [inline Pythonwin/Scintilla/License.txt:5]: Thanks for the scintilla PR, but It would help me alot if you also split this up - just keep all Scintilla stuff out of this PR for now, and we can do that in a separate PR.

File: win32/src/win32rasmodule.cpp (hunk 1 above)
Comment (raedrizqie) [reply win32/src/win32rasmodule.cpp:718]: gcc errors out because rc is not initialized.. must add 'int rc = ...' then..

Comment (mhammond) [reply win32/src/win32rasmodule.cpp:718]: That seems back to front - that line *is* initializing. I'd expect the error with `==` where there's no initialization but instead you would be evaluating an uninitialized variable.

Comment (raedrizqie) [reply win32/src/win32rasmodule.cpp]: ok, my bad.. I had reverted it back..

**Comment by raedrizqie** [comment]
Tested with:
GCC v10.2.0 (Rev3, Built by MSYS2 project)
Python v3.8.5 (default, Sep  7 2020, 18:26:56)  [GCC 10.2.0 64 bit (AMD64)]

Create a custom $(SDKDIR).
Put $(SDKDIR) in your environment.

Copy just these libs from WINDDK 7.1 to $(SDKDIR)/lib/x64:
	ADSIid.Lib
	Bits.Lib
	NtQuery.Lib
	propsys.lib

Copy just this header from WINSDK 7.1 to $(SDKDIR)/include:
	bits2_5.h : comment out '#include "bits3_0.h"'

Build...

cd pywin32-git
python setup.py build --skip-verstamp

* Why MinGw-w64 Python don't have py.exe?

*** NOTE: The following extensions were NOT built:
 exchdapi: No 64-bit library for utility functions available.
 win32ui: Unsupported due to different ABI implementations.
 win32uiole: Unsupported due to different ABI implementations.
 dde: Unsupported due to different ABI implementations.
 Pythonwin: Unsupported due to different ABI implementations.

-------------------------------------------------------------------------------

** Only applied if you want to build Pythonwin extensions. **

Create a custom $(MFCDIR) (Can be the same as your custom $(SDKDIR).
Put $(MFCDIR) in your environment.

Copy all the atlmfc headers to this custom dir.
$(MFCDIR)/atlmfc/include

Patch $(PYTHONDIR)/lib/python3.8/distutils/cygwinccompiler.py:

Find;
    def _compile(self, obj, src, ext, cc_args, extra_postargs, pp_opts):
        """Compiles the source by spawning GCC and windres if needed."""
        if ext == '.rc' or ext == '.res':
            # gcc needs '.res' and '.rc' compiled to object files !!!
            try:
                self.spawn(["windres", "-i", src, "-o", obj]) <-- this line
            except DistutilsExecError as msg:
                raise CompileError(msg)

change to;
                self.spawn(["windres", "-i", src, "-o", obj] + pp_opts)

**Comment by joankaradimov** [comment]
@raedrizqie The [scintialla pull request](https://github.com/mirror/scintilla/pull/10) you've opened is in a mirror repo. I don't think any of the maintainers are looking at it.

May I suggest you try and involve the maintainers over at https://sourceforge.net/projects/scintilla/. They appear to have a bug tracker [over there](https://sourceforge.net/p/scintilla/feature-requests/).

**Comment by raedrizqie** [comment]
Hi Joan,
Can we chat in private? I want to discuss something with you.

**Comment by mhammond** [comment]
It looks like this has stalled - feel free to reopen this if you want to update it and we can discuss some of the changes in more detail
